bin/propolis-server/src/lib/vm/ensure.rs

```diff
@@ -0,0 +1,359 @@
+            guard.resume_kernel_vm();
+            self.kernel_vm_paused = false;
+        }
+
+        self.state_publisher
+            .update(ExternalStateUpdate::Instance(InstanceState::Failed));
+
+        // Since there are extant VM objects, move to the Rundown state. The VM
+        // will move to RundownComplete when the objects are finally dropped.
+        self.vm.set_rundown().await;
+    }
+
+    /// Yields the VM objects and input queue for this VM so that they can be
+    /// used to start a state driver loop.
+    pub(super) fn into_inner(self) -> (Arc<VmObjects>, Arc<InputQueue>) {
+        (self.vm_objects, self.input_queue)
+    }
+}
+
+/// Initializes a set of Propolis components from the supplied instance spec.
+async fn initialize_vm_objects_from_spec(
+    log: &slog::Logger,
+    event_queue: &Arc<InputQueue>,
+    properties: &InstanceProperties,
+    spec: &VersionedInstanceSpec,
+    options: &EnsureOptions,
+) -> anyhow::Result<InputVmObjects> {
+    info!(log, "initializing new VM";
+              "spec" => #?spec,
+              "properties" => #?properties,
+              "use_reservoir" => options.use_reservoir,
+              "bootrom" => %options.toml_config.bootrom.display());
+
+    let vmm_log = log.new(slog::o!("component" => "vmm"));
+
+    // Set up the 'shell' instance into which the rest of this routine will
+    // add components.
+    let VersionedInstanceSpec::V0(v0_spec) = &spec;
+    let machine = build_instance(
+        &properties.vm_name(),
+        v0_spec,
+        options.use_reservoir,
+        vmm_log,
+    )?;
+
+    let mut init = MachineInitializer {
+        log: log.clone(),
+        machine: &machine,
+        devices: Default::default(),
+        block_backends: Default::default(),
+        crucible_backends: Default::default(),
+        spec: v0_spec,
+        properties,
+        toml_config: &options.toml_config,
+        producer_registry: options.oximeter_registry.clone(),
+        state: MachineInitializerState::default(),
+    };
+
+    init.initialize_rom(options.toml_config.bootrom.as_path())?;
+    let chipset = init.initialize_chipset(
+        &(event_queue.clone()
+            as Arc<dyn super::guest_event::ChipsetEventHandler>),
+    )?;
+
+    init.initialize_rtc(&chipset)?;
+    init.initialize_hpet()?;
+
+    let com1 = Arc::new(init.initialize_uart(&chipset)?);
+    let ps2ctrl = init.initialize_ps2(&chipset)?;
+    init.initialize_qemu_debug_port()?;
+    init.initialize_qemu_pvpanic(properties.into())?;
+    init.initialize_network_devices(&chipset)?;
+
+    #[cfg(not(feature = "omicron-build"))]
+    init.initialize_test_devices(&options.toml_config.devices)?;
+    #[cfg(feature = "omicron-build")]
+    info!(log, "`omicron-build` feature enabled, ignoring any test devices");
+
+    #[cfg(feature = "falcon")]
+    init.initialize_softnpu_ports(&chipset)?;
+    #[cfg(feature = "falcon")]
+    init.initialize_9pfs(&chipset)?;
```

Comment:
unimportant style nit, take it or leave it: why not
```suggestion
    #[cfg(feature = "falcon")] {
        init.initialize_softnpu_ports(&chipset)?;
        init.initialize_9pfs(&chipset)?;
    }
```